substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -40,8 +41,60 @@ use sp_std::{fmt, prelude::*};
 /// the decoding fails.
 const EXTRINSIC_FORMAT_VERSION: u8 = 4;
 
-/// The `SingaturePayload` of `UncheckedExtrinsic`.
-type UncheckedSignaturePayload<Address, Signature, Extra> = (Address, Signature, Extra);
+/// The `SignaturePayload` of `UncheckedExtrinsic`.
+type UncheckedSignaturePayload<Address, Signature, Extension> = (Address, Signature, Extension);
+
+impl<Address: TypeInfo, Signature: TypeInfo, Extension: TypeInfo> SignaturePayload
+	for UncheckedSignaturePayload<Address, Signature, Extension>
+{
+	type SignatureAddress = Address;
+	type Signature = Signature;
+	type SignatureExtra = Extension;
+}
+
+/// A "header" for extrinsics leading up to the call itself. Determines the type of extrinsic and
+/// holds any necessary specialized data.
+#[derive(Eq, PartialEq, Clone, Encode, Decode)]
+pub enum Preamble<Address, Signature, Extension> {
+	/// An extrinsic without a signature or any extension. This means it's either an inherent or
+	/// an old-school "Unsigned" (we don't use that terminology any more since it's confusable with
+	/// the general transaction which is without a signature but does have an extension).
+	///
+	/// NOTE: In the future, once we remove `ValidateUnsigned`, this will only serve Inherent
+	/// extrinsics and thus can be renamed to `Inherent`.
+	#[codec(index = 0b00000100)]
+	Bare,
+	/// An old-school transaction extrinsic which includes a signature of some hard-coded crypto.
+	#[codec(index = 0b10000100)]
+	Signed(Address, Signature, Extension),
+	/// A new-school transaction extrinsic which does not include a signature.
+	#[codec(index = 0b01000100)]
```

Comment:
Previously we would decode the version and "is signed" from this index like so:

```rust
let is_signed = version & 0b1000_0000 != 0;
let version = version & 0b0111_1111;
```

Using that approach, the `version` here for `General` is now 68. I wonder whether there is a strong reason for this and whether the `General` index should instead be `0b00000101` (ie increment the version to 5) instead?

We also still have `const EXTRINSIC_FORMAT_VERSION: u8 = 4;` further up, which feels a bit odd if there are now two versions in play in this file; not sure what best to do there (should we add another constant for the new version, or rename/remove this one?)

Reply:
I view `EXTRINSIC_FORMAT_VERSION` as the major version in a semver sense, which means that we would bump it if we make backwards incompatible changes (i.e. decoding would fail with the new format). This is not the case in the implementation presented in this PR.

However, now identifying the correct format requires 2 (8th and 7th) bits instead of 1 (8th), as we now deal with 3 choices instead of 2 (see `Preamble`). The 8th and 7th bits will identify the type of extrinsic, the rest of them identify the extrinsic format version. It's fine since we didn't reach a format version of `64` or greater before this change, so that 7th bit was unused.

Reply:
If I were designing this from scratch nowadays I'd consider having _all_ of the bits represent the format version; different bits mean you need to decode it differently. I wouldn't bother having a split between "is it signed or not" and "this is the version" (I'm not sure what you gain from that now).

Anyway, from the POV of external libraries what encode/decode extrinsics, if a library previously checked the 7 trailing bits for the version, it will quickly see that it can't decode this new `General` extrinsic anyway (it'll appear to have a version of 68), so this all works out anyway :)